### PR TITLE
Loss function as an object

### PR DESCRIPTION
**Pull request by alexhernandezgarcia**
This PR refactors the code around the GFlowNet loss functions to make the loss function a class and each loss function a separate class.

This change is something we had wanted to do for a while, but it was never a high priority since the impact is rather small. However, since we may implement a couple of new loss functions soon, I thought it would be worth making the change now. 

This is a more modular implementation of the loss function, which makes the code in the GFlowNet agent a bit cleaner and the implementation of new loss functions hopefully clearer.

### To be done in this PR
- [x] Adapt experiment config files
- [x] Remove optimizer.loss from all gflownet configs, e.g. here: https://github.com/alexhernandezgarcia/gflownet/blob/a55ffda346d6c9804b3cf31159eea2c5e8dfac35/config/gflownet/detailedbalance.yaml#L6
- [x] Clean up part about moving average of the loss
- [x] Sanity checks
- [x] Make current tests pass

### To be done in future PRs
- [ ] Make the optimiser modular too
- [ ] Improve the policies part of the code to remove the dependence on the loss function (see https://github.com/alexhernandezgarcia/gflownet/pull/327)
- [ ] Instructions to create new loss functions
- [ ] Unit tests for the losses

### CODE REVIEW

File: gflownet/gflownet.py [hunk 1]
```diff
@@ -130,34 +135,20 @@ def __init__(
...
-            )
         # Loss
-        if optimizer.loss in ["flowmatch", "flowmatching"]:
-            self.loss = "flowmatch"
-            self.logZ = None
-        elif optimizer.loss in ["trajectorybalance", "tb"]:
-            self.loss = "trajectorybalance"
+        self.loss = loss
+        if self.loss.requires_log_z:
             self.logZ = nn.Parameter(torch.ones(optimizer.z_dim) * 150.0 / 64)
```
Comment (AlexandraVolokhova) [inline gflownet/gflownet.py:141]: I've never noticed this before, but what is there a reason for this multiplication `* 150.0 / 64` ?

File: gflownet/gflownet.py [hunk 2]
```diff
@@ -1372,8 +1059,9 @@ def log_train_iteration(self, pbar: tqdm, losses: List, batch: Batch, times: dic
             )
 
             # Log losses
+            losses["Loss"] = losses["all"]
```
Comment (AlexandraVolokhova) [inline gflownet/gflownet.py:1062]: I think you need to delete the key "all" in the dictionary `losses` as well, otherwise, you're logging the same metric twice under the names "Loss" and "all"

File: gflownet/losses/__init__.py
Comment (AlexandraVolokhova) [inline gflownet/losses/__init__.py:1]: I'm confused with this file, could you elaborate why it is needed? We haven't created __init__.py for other subfolders like env and policy, why do we need it for losses? The docstring inside is not very clear to me

File: gflownet/utils/batch.py [hunk 3]
```diff
@@ -1628,3 +1629,77 @@ def get_item(
                 "item must be one of: state, parent, action, done, mask_f[orward] or "
                 "mask_b[ackward]"
             )
+
+
+def compute_logprobs_trajectories(
+    batch: Batch,
+    env: GFlowNetEnv,
```
Comment (AlexandraVolokhova) [inline gflownet/utils/batch.py:1636]: I'm thinking that maybe passing `env` is not necessary here because we can always access` batch.readonly_env` if we need some methods of the env. And if we remove dependency on the env hre, we can also remove it in the Loss class (if I understand correctly, `self.env` is used there only to call `compute_logprobs_trajectories` function)

File: gflownet/losses/__init__.py
Comment (alexhernandezgarcia) [reply gflownet/losses/__init__.py:1]: I'm confused because the `__init__.py` file also is there for the envs and policy (and other modules) because it is (was?) required. I have actually just learnt that it may not be required any longer in recent versions of Python.

https://stackoverflow.com/questions/448271/what-is-init-py-for

File: gflownet/gflownet.py (hunk 1 above)
Comment (AlexandraVolokhova) [reply gflownet/gflownet.py:141]: I'd put a TODO item to make this constant a parameter of the training

File: gflownet/utils/batch.py (hunk 3 above)
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py:1636]: But we can also use `batch.readonly_env` inside the detailed balance loss call, can't we?

File: gflownet/losses/base.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+from gflownet.utils.common import set_device, set_float_precision
+
+
+class BaseLoss(metaclass=ABCMeta):
+    # TODO: study whether the dependence on the environment can be removed.
+    # TODO: improve dependence on policies (needs re-implementation of policies)
+    def __init__(
+        self,
+        env_maker: partial,
+        forward_policy: Policy,
```
Comment (LenaNehale) [inline gflownet/losses/base.py:29]: I am still getting familiar with the codebase so correct me if I am wrong! 
I was confused to see forward_policy used in the flow-matching loss, because there we are learning flows (different from forward policy, which implies that the distribution is normalized to sum to 1), what do you think of adding another parameter "state_action_flow: Policy" as using this for flow matching? It's just for naming purposes, otherwise the code for flow-matching makes sense

Comment (alexhernandezgarcia) [reply gflownet/losses/base.py:29]: I see your point but I think it would be more confusing if we end up having different names for the same thing. The meaning of "policy" in the codebase is "a model used to learn a sampling policy". It is valid for the flow matching because these policy models do not learn a (normalised) probability distribution, so their outputs can be interpreted as flows. And we can use the same models and notion for other losses, such as TB, because we can the outputs of these policy models and normalise them to obtain a probability.

Does that make sense?

Comment (LenaNehale) [reply gflownet/losses/base.py:29]: Oh ok, it makes sense, thank you!

File: gflownet/utils/batch.py (hunk 3 above)
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py]: Nice, looks good to me!

File: gflownet/losses/trajectorybalance.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            batch, forward_policy=self.forward_policy, backward=False
+        )
+        logprobs_b = compute_logprobs_trajectories(
+            batch, backward_policy=self.backward_policy, backward=True
+        )
+        # Get rewards from batch
+        logrewards = batch.get_terminating_rewards(log=True, sort_by="trajectory")
+
+        # Trajectory balance loss
+        return (self.logZ.sum() + logprobs_f - logprobs_b - logrewards).pow(2)
```
Comment (LenaNehale) [inline gflownet/losses/trajectorybalance.py:111]: I just saw this, I think it should be logZ.mean() , not logZ.sum()  , otherwise the final logZ changes with the batch size. If the batch size is fixed during training it's fine, but I would change it to avoid problems in case the bs changes (e.g. in the case of training with a replay buffer, at the beginning we still don't have enough samples in the replay buffer so the actual bs is smaller)
